flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java

```diff
@@ -902,6 +912,39 @@ private void deploySafely(ExecutionVertex executionVertex) {
         private void handleDeploymentFailure(ExecutionVertex executionVertex, JobException e) {
             executionVertex.markFailed(e);
         }
+
+        @Override
+        public void newResourcesAvailable() {
+            int availableSlots = declarativeSlotPool.getFreeSlotsInformation().size();
+
+            if (availableSlots > 0) {
+                LOG.info(
+                        "{} additional slots present. Checking if rescaling using all slots would increase parallelism",
+                        availableSlots);
+
+                // check if rescaling would increase the parallelism of at least one vertex
+                final Optional<? extends VertexAssignment> potentialNewParallelism =
+                        slotAllocator.determineParallelism(
+                                new JobGraphJobInformation(jobGraph),
+                                declarativeSlotPool.getAllSlotsInformation());
+
+                if (potentialNewParallelism.isPresent()
+                        && isParallelismHigher(potentialNewParallelism.get())) {
+                    transitionToState(new Restarting(executionGraph, 0L));
```

Comment:
Something should be logged along the lines of "restarting the job to scale up". It could also be interesting to log the cumulative different between the old and new parallelism.